vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphiQLHandlerImpl.java

```diff
@@ -0,0 +1,152 @@
+import io.vertx.core.http.impl.MimeMapping;
+import io.vertx.core.impl.logging.Logger;
+import io.vertx.core.impl.logging.LoggerFactory;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.MIMEHeader;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.graphql.GraphiQLHandler;
+import io.vertx.ext.web.handler.graphql.GraphiQLHandlerOptions;
+import io.vertx.ext.web.impl.Utils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Scanner;
+import java.util.function.Function;
+
+/**
+ * @author Thomas Segismont
+ */
+public class GraphiQLHandlerImpl implements GraphiQLHandler {
+
+  private static final Logger log = LoggerFactory.getLogger(GraphiQLHandlerImpl.class);
+
+  private static final Function<RoutingContext, MultiMap> DEFAULT_GRAPHIQL_REQUEST_HEADERS_FACTORY = rc -> null;
+
+  private final GraphiQLHandlerOptions options;
+
+  private Function<RoutingContext, MultiMap> graphiQLRequestHeadersFactory = DEFAULT_GRAPHIQL_REQUEST_HEADERS_FACTORY;
+
+  public GraphiQLHandlerImpl(GraphiQLHandlerOptions options) {
+    Objects.requireNonNull(options, "options");
+    this.options = options;
+  }
+
+  @Override
+  public GraphiQLHandler graphiQLRequestHeaders(Function<RoutingContext, MultiMap> factory) {
+    graphiQLRequestHeadersFactory = factory != null ? factory : DEFAULT_GRAPHIQL_REQUEST_HEADERS_FACTORY;
+    return this;
+  }
+
+  @Override
+  public void handle(RoutingContext rc) {
+    if (!options.isEnabled()) {
+      rc.next();
+      return;
+    }
+    HttpServerResponse response = rc.response();
+    String filename = Utils.pathOffset(rc.normalisedPath(), rc);
+    if (filename.isEmpty()) {
+      if (rc.parsedHeaders().accept().stream().map(MIMEHeader::subComponent).anyMatch(sub -> "html".equalsIgnoreCase(sub))) {
+        rc.response().setStatusCode(301).putHeader(HttpHeaders.LOCATION, rc.currentRoute().getPath()).end();
+      } else {
+        rc.next();
+      }
+      return;
+    }
+    if (filename.equals("/")) {
+      filename = "/index.html";
+    }
+    String resource = loadResource(filename);
+    if (resource == null) {
+      rc.next();
+      return;
+    }
+    if (filename.equals("/index.html")) {
+      resource = resource.replace("__VERTX_GRAPHIQL_CONFIG__", replacement(rc));
+      response.putHeader(HttpHeaders.CACHE_CONTROL, "no-cache");
+    } else {
+      response.putHeader(HttpHeaders.CACHE_CONTROL, "max-age=31536000");
+    }
+    String contentType = MimeMapping.getMimeTypeForFilename(filename);
+    if (contentType != null) {
+      if (contentType.startsWith("text")) {
+        response.putHeader(HttpHeaders.CONTENT_TYPE, contentType + ";charset=utf8");
+      } else {
+        response.putHeader(HttpHeaders.CONTENT_TYPE, contentType);
+      }
+    }
+    response.end(resource);
+  }
+
+  private String loadResource(String relativePath) {
```

Comment:
Why not use `vertx.fileSystem()` instead? this avoids dealing with class loaders (think native images) and perhaps we could cache these files too? if they are not that many as I guess they would be called frequently (every browser refresh, etc...)

Reply:
I think I wanted this to work even when classpath resolving is disabled. But since this is for development only, it's probably fine to assume it's enabled. I'll update the code. Caching on the server side is overkill: it's for development only and the cache headers are sent for browser caching.

Reply:
Done